sql/rowexec: TestUncertaintyErrorIsReturned failed

[(sql/rowexec).TestUncertaintyErrorIsReturned failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2399268&tab=buildLog) on [master@b319a0b6858332a4d022f4c4061702fc8822db96](https://github.com/cockroachdb/cockroach/commits/b319a0b6858332a4d022f4c4061702fc8822db96):

```
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x954
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8

Goroutine 1175 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:337 +0x3e6
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:358 +0x98
  github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Start()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:109 +0x26a
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:130 +0x431
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326 +0x150
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:127 +0xcb
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:543 +0x4bc
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:505 +0xa1
  github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xbd8
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:201 +0x135
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:60 +0xbe
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:47 +0x80
  google.golang.org/grpc.chainUnaryServerInterceptors.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907 +0x110
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:507 +0x1d8
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x954
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8
==================
FAIL	github.com/cockroachdb/cockroach/pkg/sql/rowexec	16.305s

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestUncertaintyErrorIsReturned PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestUncertaintyErrorIsReturned.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/rowexec).TestUncertaintyErrorIsReturned failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2401135&tab=buildLog) on [master@70fb9a54d6c4d7b3c1d8ad93bb399eefef8f5d04](https://github.com/cockroachdb/cockroach/commits/70fb9a54d6c4d7b3c1d8ad93bb399eefef8f5d04):

```
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8

Goroutine 903 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:337 +0x3e6
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:358 +0x98
  github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Start()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:109 +0x26a
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:130 +0x431
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326 +0x150
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:127 +0xcb
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:543 +0x4bc
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:505 +0xa1
  github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xbd8
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:201 +0x135
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:60 +0xbe
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:47 +0x80
  google.golang.org/grpc.chainUnaryServerInterceptors.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907 +0x110
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:507 +0x1d8
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x954
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8
==================
        --- FAIL: TestUncertaintyErrorIsReturned/vectorize=false/SELECT_*_FROM_t_AS_t1_JOIN_t_AS_t2_ON_t1.x_=_t2.x (0.13s)
            testing.go:853: race detected during execution of test
=== RUN   TestUncertaintyErrorIsReturned/vectorize=false
    --- FAIL: TestUncertaintyErrorIsReturned/vectorize=false (0.34s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestUncertaintyErrorIsReturned PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestUncertaintyErrorIsReturned.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/rowexec/processors_test.go
+++ b/pkg/sql/rowexec/processors_test.go
@@ -701,7 +701,6 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 // in which an UncertaintyError is expected.
 func TestUncertaintyErrorIsReturned(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.UnderRaceWithIssue(t, 56027, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	const numNodes = 3

--- a/pkg/sql/rowflow/routers.go
+++ b/pkg/sql/rowflow/routers.go
@@ -114,6 +114,9 @@ type routerOutput struct {
 	// infrastructure of rowBufToPushFrom.
 	rowBufToPushFromMon *mon.BytesMonitor
 	rowBufToPushFromAcc *mon.BoundAccount
+	// rowBufToPushFromRowSize stores the size of the row that we have
+	// accounted for when adding it to rowBufToPushFrom buffer in ith position.
+	rowBufToPushFromRowSize [routerRowBufSize]int64
 }
 
 func (ro *routerOutput) addMetadataLocked(meta *execinfrapb.ProducerMetadata) {
@@ -152,11 +155,10 @@ func (ro *routerOutput) popRowsLocked(ctx context.Context) ([]rowenc.EncDatumRow
 		// We're reusing the same rowBufToPushFrom slice, so we can only
 		// release the memory under the "old" row once we overwrite it in
 		// rowBufToPushFrom which we're about to do for rowBufToPushFrom[n].
-		prevSize := uintptr(0)
-		if ro.rowBufToPushFrom[n] != nil {
-			prevSize = ro.rowBufToPushFrom[n].Size()
-		}
-		if err := ro.rowBufToPushFromAcc.Grow(ctx, int64(row.Size()-prevSize)); err != nil {
+		rowSize := int64(row.Size())
+		delta := rowSize - ro.rowBufToPushFromRowSize[n]
+		ro.rowBufToPushFromRowSize[n] = rowSize
+		if err := ro.rowBufToPushFromAcc.Grow(ctx, delta); err != nil {
 			return err
 		}
 		ro.rowBufToPushFrom[n] = row

